include/boost/rational.hpp

```diff
@@ -148,33 +152,102 @@ class rational :
-    rational& assign(param_type n, param_type d);
+    template <class T>
+    typename enable_if_c<
+       rational_detail::is_compatible_integer<T, IntType>::value, rational &
+    >::type assign(const T& n, const T& d)
+    {
+       return *this = rational<IntType>(static_cast<IntType>(n), static_cast<IntType>(d));
+    }
+    //
+    // The following overloads should probably *not* be provided - 
+    // but are provided for backwards compatibity reasons only.
+    // These allow for construction/assignment from types that
+    // are wider than IntType only if there is an implicit
+    // conversion from T to IntType, they will throw a bad_rational
+    // if the conversion results in loss of precision or undefined behaviour.
+    //
+    template <class T>
+    rational(const T& n, typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value
+    >::type const* = 0)
+    {
+       assign(n, static_cast<T>(1));
+    }
+    template <class T>
+    rational(const T& n, const T& d, typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value
+    >::type const* = 0)
+    {
+       assign(n, d);
+    }
+    template <class T>
+    typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value,
+       rational &
+    >::type operator=(const T& n) { return assign(n, static_cast<T>(1)); }
+
+    template <class T>
+    typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value,
+       rational &
+    >::type assign(const T& n, const T& d)
+    {
+       if(!is_safe_narrowing_conversion(n) || !is_safe_narrowing_conversion(d))
+          BOOST_THROW_EXCEPTION(bad_rational());
```

Comment:
I agree that the same reasoning applies. The only two sensible behaviors I see for op+-*/ on overflow are to either throw, or return the closest representable. (For op- with an underlying unsigned type, throwing when the result would be negative looks like the only option.)

This does present a bit of a problem with op++/-- though, because an overflow/underflow check there would be measurable, so it should probably be skipped. The special case of +- with the same denominator would also be slowed down by the check.

I agree that this is not a trivial decision.